File: trunk/symmetric/symmetric-core/src/main/java/org/jumpmind/symmetric/service/impl/RouterService.java
/*
 * Licensed to JumpMind Inc under one or more contributor 
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding 
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU Lesser General Public License (the
 * "License"); you may not use this file except in compliance
 * with the License. 
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, see           
 * <http://www.gnu.org/licenses/>.
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License. 
 */
package org.jumpmind.symmetric.service.impl;

import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.atomic.AtomicInteger;

import org.apache.commons.lang.StringUtils;
import org.jumpmind.db.model.Table;
import org.jumpmind.db.sql.AbstractSqlMap;
import org.jumpmind.symmetric.common.Constants;
import org.jumpmind.symmetric.common.ParameterConstants;
import org.jumpmind.symmetric.db.ISymmetricDialect;
import org.jumpmind.symmetric.model.Data;
import org.jumpmind.symmetric.model.DataMetaData;
import org.jumpmind.symmetric.model.Node;
import org.jumpmind.symmetric.model.NodeChannel;
import org.jumpmind.symmetric.model.NodeGroupLink;
import org.jumpmind.symmetric.model.NodeSecurity;
import org.jumpmind.symmetric.model.OutgoingBatch;
import org.jumpmind.symmetric.model.OutgoingBatch.Status;
import org.jumpmind.symmetric.model.Router;
import org.jumpmind.symmetric.model.TriggerRouter;
import org.jumpmind.symmetric.route.BshDataRouter;
import org.jumpmind.symmetric.route.ChannelRouterContext;
import org.jumpmind.symmetric.route.ColumnMatchDataRouter;
import org.jumpmind.symmetric.route.ConfigurationChangedDataRouter;
import org.jumpmind.symmetric.route.DataGapDetector;
import org.jumpmind.symmetric.route.DataGapRouteReader;
import org.jumpmind.symmetric.route.DefaultBatchAlgorithm;
import org.jumpmind.symmetric.route.DefaultDataRouter;
import org.jumpmind.symmetric.route.IBatchAlgorithm;
import org.jumpmind.symmetric.route.IDataRouter;
import org.jumpmind.symmetric.route.IDataToRouteGapDetector;
import org.jumpmind.symmetric.route.IDataToRouteReader;
import org.jumpmind.symmetric.route.LookupTableDataRouter;
import org.jumpmind.symmetric.route.NonTransactionalBatchAlgorithm;
import org.jumpmind.symmetric.route.SimpleRouterContext;
import org.jumpmind.symmetric.route.SubSelectDataRouter;
import org.jumpmind.symmetric.route.TransactionalBatchAlgorithm;
import org.jumpmind.symmetric.service.ClusterConstants;
import org.jumpmind.symmetric.service.IClusterService;
import org.jumpmind.symmetric.service.IConfigurationService;
import org.jumpmind.symmetric.service.IDataService;
import org.jumpmind.symmetric.service.INodeService;
import org.jumpmind.symmetric.service.IOutgoingBatchService;
import org.jumpmind.symmetric.service.IParameterService;
import org.jumpmind.symmetric.service.IRouterService;
import org.jumpmind.symmetric.service.ITransformService;
import org.jumpmind.symmetric.service.ITriggerRouterService;
import org.jumpmind.symmetric.statistic.IStatisticManager;
import org.jumpmind.symmetric.statistic.StatisticConstants;

/**
 * @see IRouterService
 */
public class RouterService extends AbstractService implements IRouterService {

    private IClusterService clusterService;

    private IDataService dataService;

    private IConfigurationService configurationService;

    private ITriggerRouterService triggerRouterService;

    private IOutgoingBatchService outgoingBatchService;

    private INodeService nodeService;

    private Map<String, IDataRouter> routers;

    private Map<String, IBatchAlgorithm> batchAlgorithms;

    private IStatisticManager statisticManager;

    transient ExecutorService readThread = null;

    public RouterService(IParameterService parameterService, ISymmetricDialect symmetricDialect,
            IClusterService clusterService, IDataService dataService,
            IConfigurationService configurationService, ITriggerRouterService triggerRouterService,
            IOutgoingBatchService outgoingBatchService, INodeService nodeService,
            IStatisticManager statisticManager, ITransformService transformService) {
        super(parameterService, symmetricDialect);
        this.clusterService = clusterService;
        this.dataService = dataService;
        this.configurationService = configurationService;
        this.triggerRouterService = triggerRouterService;
        this.outgoingBatchService = outgoingBatchService;
        this.nodeService = nodeService;
        this.statisticManager = statisticManager;

        this.batchAlgorithms = new HashMap<String, IBatchAlgorithm>();
        this.batchAlgorithms.put("default", new DefaultBatchAlgorithm());
        this.batchAlgorithms.put("nontransactional", new NonTransactionalBatchAlgorithm());
        this.batchAlgorithms.put("transactional", new TransactionalBatchAlgorithm());

        this.routers = new HashMap<String, IDataRouter>();
        this.routers.put("configurationChanged", new ConfigurationChangedDataRouter(
                configurationService, nodeService, triggerRouterService, parameterService,
                transformService));
        this.routers.put("bsh", new BshDataRouter(symmetricDialect));
        this.routers.put("subselect", new SubSelectDataRouter(symmetricDialect));
        this.routers.put("lookuptable", new LookupTableDataRouter(symmetricDialect));
        this.routers.put("default", new DefaultDataRouter());
        this.routers.put("column", new ColumnMatchDataRouter(configurationService));
    }

    /**
     * For use in data load events
     */
    public boolean shouldDataBeRouted(SimpleRouterContext context, DataMetaData dataMetaData,
            Set<Node> nodes, boolean initialLoad) {
        IDataRouter router = getDataRouter(dataMetaData.getTriggerRouter());
        Collection<String> nodeIds = router.routeToNodes(context, dataMetaData, nodes, initialLoad);
        for (Node node : nodes) {
            if (nodeIds != null && nodeIds.contains(node.getNodeId())) {
                return true;
            }
        }
        return false;
    }

    protected synchronized ExecutorService getReadService() {
        if (readThread == null) {
            readThread = Executors.newCachedThreadPool(new ThreadFactory() {
                final AtomicInteger threadNumber = new AtomicInteger(1);
                final String namePrefix = parameterService.getEngineName().toLowerCase()
                        + "-router-reader-";

                public Thread newThread(Runnable r) {
                    Thread t = new Thread(r);
                    t.setName(namePrefix + threadNumber.getAndIncrement());
                    if (t.isDaemon()) {
                        t.setDaemon(false);
                    }
                    if (t.getPriority() != Thread.NORM_PRIORITY) {
                        t.setPriority(Thread.NORM_PRIORITY);
                    }
                    return t;
                }
            });
        }
        return readThread;
    }

    public synchronized void stop() {
        try {
            log.info("RouterShuttingDown");
            getReadService().shutdown();
            readThread = null;
        } catch (Exception ex) {
            log.error(ex);
        }
    }

    public synchronized void destroy() {

    }

    /**
     * This method will route data to specific nodes.
     */
    synchronized public long routeData() {
        long dataCount = -1l;
        if (clusterService.lock(ClusterConstants.ROUTE)) {
            try {
                insertInitialLoadEvents();
                long ts = System.currentTimeMillis();
                IDataToRouteGapDetector gapDetector = new DataGapDetector(log, dataService,
                        parameterService, symmetricDialect, getSqlMap());
                gapDetector.beforeRouting();
                dataCount = routeDataForEachChannel();
                gapDetector.afterRouting();
                ts = System.currentTimeMillis() - ts;
                if (dataCount > 0 || ts > Constants.LONG_OPERATION_THRESHOLD) {
                    log.info("RoutedDataInTime", dataCount, ts);
                }
            } finally {
                clusterService.unlock(ClusterConstants.ROUTE);
            }
        }
        return dataCount;
    }

    protected void insertInitialLoadEvents() {
        try {
            Node identity = nodeService.findIdentity();
            if (identity != null) {
                Map<String, NodeSecurity> nodeSecurities = nodeService.findAllNodeSecurity(false);
                if (nodeSecurities != null) {
                    for (NodeSecurity security : nodeSecurities.values()) {
                        if (!security.getNodeId().equals(identity.getNodeId())
                                && security.isInitialLoadEnabled()
                                && (security.getRegistrationTime() != null || security.getNodeId()
                                        .equals(identity.getCreatedAtNodeId()))) {
                            long ts = System.currentTimeMillis();
                            dataService.insertReloadEvents(nodeService.findNode(security
                                    .getNodeId()));
                            ts = System.currentTimeMillis() - ts;
                            if (ts > Constants.LONG_OPERATION_THRESHOLD) {
                                log.warn("ReloadedNode", security.getNodeId(), ts);
                            } else {
                                log.info("ReloadedNode", security.getNodeId(), ts);
                            }
                        }
                    }
                }
            }
        } catch (Exception ex) {
            log.error(ex);
        }
    }

    /**
     * We route data channel by channel for two reasons. One is that if/when we
     * decide to multi-thread the routing it is a simple matter of inserting a
     * thread pool here and waiting for all channels to be processed. The other
     * reason is to reduce the number of connections we are required to have.
     */
    protected int routeDataForEachChannel() {
        Node sourceNode = nodeService.findIdentity();
        final List<NodeChannel> channels = configurationService.getNodeChannels(false);
        int dataCount = 0;
        for (NodeChannel nodeChannel : channels) {
            if (!nodeChannel.isSuspendEnabled() && nodeChannel.isEnabled()) {
                dataCount += routeDataForChannel(nodeChannel, sourceNode);
            } else {
                if (log.isDebugEnabled()) {
                    log.debug("RouterSkippingChannel", nodeChannel.getChannelId());
                }
            }
        }
        return dataCount;
    }

    protected int routeDataForChannel(final NodeChannel nodeChannel, final Node sourceNode) {
        ChannelRouterContext context = null;
        long ts = System.currentTimeMillis();
        int dataCount = -1;
        try {
            context = new ChannelRouterContext(sourceNode.getNodeId(), nodeChannel,
                    symmetricDialect.getPlatform().getSqlTemplate().startSqlTransaction());
            dataCount = selectDataAndRoute(context);
            return dataCount;
        } catch (Exception ex) {
            log.error("RouterRoutingFailed", ex, nodeChannel.getChannelId());
            if (context != null) {
                context.rollback();
            }
            return 0;
        } finally {
            try {
                if (dataCount > 0) {
                    long insertTs = System.currentTimeMillis();
                    dataService.insertDataEvents(context.getSqlTransaction(),
                            context.getDataEventList());
                    context.clearDataEventsList();
                    completeBatchesAndCommit(context);
                    context.incrementStat(System.currentTimeMillis() - insertTs,
                            ChannelRouterContext.STAT_INSERT_DATA_EVENTS_MS);
                    if (context.getLastDataIdProcessed() > 0) {
                        String channelId = nodeChannel.getChannelId();
                        long queryTs = System.currentTimeMillis();
                        long dataLeftToRoute = sqlTemplate.queryForInt(
                                getSql("selectUnroutedCountForChannelSql"), channelId,
                                context.getLastDataIdProcessed());
                        queryTs = System.currentTimeMillis() - queryTs;
                        if (queryTs > Constants.LONG_OPERATION_THRESHOLD) {
                            log.warn("UnRoutedQueryTookLongTime", channelId, queryTs);
                        }
                        statisticManager.setDataUnRouted(channelId, dataLeftToRoute);
                    }
                }
            } catch (Exception e) {
                if (context != null) {
                    context.rollback();
                }
                log.error(e);
            } finally {
                long totalTime = System.currentTimeMillis() - ts;
                context.incrementStat(totalTime, ChannelRouterContext.STAT_ROUTE_TOTAL_TIME);
                context.logStats(log, totalTime);
                context.cleanup();
            }
        }
    }

    protected void completeBatchesAndCommit(ChannelRouterContext context) throws SQLException {
        Set<IDataRouter> usedRouters = new HashSet<IDataRouter>(context.getUsedDataRouters());
        List<OutgoingBatch> batches = new ArrayList<OutgoingBatch>(context.getBatchesByNodes()
                .values());
        context.commit();
        for (OutgoingBatch batch : batches) {
            batch.setRouterMillis(System.currentTimeMillis() - batch.getCreateTime().getTime());
            for (IDataRouter dataRouter : usedRouters) {
                dataRouter.completeBatch(context, batch);
            }
            if (Constants.UNROUTED_NODE_ID.equals(batch.getNodeId())) {
                batch.setStatus(Status.OK);
            } else {
                batch.setStatus(Status.NE);
            }
            outgoingBatchService.updateOutgoingBatch(batch);
            context.getBatchesByNodes().remove(batch.getNodeId());
        }

        for (IDataRouter dataRouter : usedRouters) {
            dataRouter.contextCommitted(context);
        }
        context.setNeedsCommitted(false);
    }

    protected Set<Node> findAvailableNodes(TriggerRouter triggerRouter, ChannelRouterContext context) {
        Set<Node> nodes = context.getAvailableNodes().get(triggerRouter);
        if (nodes == null) {
            nodes = new HashSet<Node>();
            Router router = triggerRouter.getRouter();
            NodeGroupLink link = configurationService.getNodeGroupLinkFor(router.getNodeGroupLink()
                    .getSourceNodeGroupId(), router.getNodeGroupLink().getTargetNodeGroupId());
            if (link != null) {
                nodes.addAll(nodeService.findEnabledNodesFromNodeGroup(router.getNodeGroupLink()
                        .getTargetNodeGroupId()));
            } else {
                log.error("RouterIllegalNodeGroupLink", router.getRouterId(), router
                        .getNodeGroupLink().getSourceNodeGroupId(), router.getNodeGroupLink()
                        .getTargetNodeGroupId());
            }
            context.getAvailableNodes().put(triggerRouter, nodes);
        }
        return nodes;
    }

    /**
     * Pre-read data and fill up a queue so we can peek ahead to see if we have
     * crossed a database transaction boundary. Then route each {@link Data}
     * while continuing to keep the queue filled until the result set is
     * entirely read.
     * 
     * @param conn
     *            The connection to use for selecting the data.
     * @param context
     *            The current context of the routing process
     */
    protected int selectDataAndRoute(ChannelRouterContext context) throws SQLException {
        IDataToRouteReader reader = new DataGapRouteReader(log, getSqlMap(), context, dataService,
                symmetricDialect, parameterService);
        getReadService().execute(reader);
        Data data = null;
        int totalDataCount = 0;
        int totalDataEventCount = 0;
        int statsDataCount = 0;
        int statsDataEventCount = 0;
        final int maxNumberOfEventsBeforeFlush = parameterService
                .getInt(ParameterConstants.ROUTING_FLUSH_JDBC_BATCH_SIZE);
        try {
            do {
                data = reader.take();
                if (data != null) {
                    context.setLastDataIdProcessed(data.getDataId());
                    statsDataCount++;
                    totalDataCount++;
                    int dataEventsInserted = routeData(data, context);
                    statsDataEventCount += dataEventsInserted;
                    totalDataEventCount += dataEventsInserted;
                    long insertTs = System.currentTimeMillis();
                    try {
                        if (maxNumberOfEventsBeforeFlush <= context.getDataEventList().size()
                                || context.isNeedsCommitted()) {
                            dataService.insertDataEvents(context.getSqlTransaction(),
                                    context.getDataEventList());
                            context.clearDataEventsList();
                        }
                        if (context.isNeedsCommitted()) {
                            completeBatchesAndCommit(context);
                        }
                    } finally {
                        context.incrementStat(System.currentTimeMillis() - insertTs,
                                ChannelRouterContext.STAT_INSERT_DATA_EVENTS_MS);

                        if (statsDataCount > StatisticConstants.FLUSH_SIZE_ROUTER_DATA) {
                            statisticManager.incrementDataRouted(context.getChannel()
                                    .getChannelId(), statsDataCount);
                            statsDataCount = 0;
                            statisticManager.incrementDataEventInserted(context.getChannel()
                                    .getChannelId(), statsDataEventCount);
                            statsDataEventCount = 0;
                        }
                    }
                }
            } while (data != null);

        } finally {
            reader.setReading(false);
            if (statsDataCount > 0) {
                statisticManager.incrementDataRouted(context.getChannel().getChannelId(),
                        statsDataCount);
            }
            if (statsDataEventCount > 0) {
                statisticManager.incrementDataEventInserted(context.getChannel().getChannelId(),
                        statsDataEventCount);
            }
        }
        context.incrementStat(totalDataCount, ChannelRouterContext.STAT_DATA_ROUTED_COUNT);
        return totalDataEventCount;

    }

    protected int routeData(Data data, ChannelRouterContext context) throws SQLException {
        int numberOfDataEventsInserted = 0;
        context.recordTransactionBoundaryEncountered(data);
        List<TriggerRouter> triggerRouters = getTriggerRoutersForData(data);
        if (triggerRouters != null && triggerRouters.size() > 0) {
            for (TriggerRouter triggerRouter : triggerRouters) {
                Table table = symmetricDialect.getTable(triggerRouter.getTrigger(), true);
                DataMetaData dataMetaData = new DataMetaData(data, table, triggerRouter,
                        context.getChannel());
                Collection<String> nodeIds = null;
                if (!context.getChannel().isIgnoreEnabled()
                        && triggerRouter.isRouted(data.getDataEventType())) {
                    IDataRouter dataRouter = getDataRouter(triggerRouter);
                    context.addUsedDataRouter(dataRouter);
                    long ts = System.currentTimeMillis();
                    nodeIds = dataRouter.routeToNodes(context, dataMetaData,
                            findAvailableNodes(triggerRouter, context), false);
                    context.incrementStat(System.currentTimeMillis() - ts,
                            ChannelRouterContext.STAT_DATA_ROUTER_MS);

                    if (!triggerRouter.isPingBackEnabled() && data.getSourceNodeId() != null
                            && nodeIds != null) {
                        nodeIds.remove(data.getSourceNodeId());
                    }
                }

                numberOfDataEventsInserted += insertDataEvents(context, dataMetaData, nodeIds,
                        triggerRouter);
            }

        } else {
            log.warn("TriggerProcessingFailedMissing", data.getTriggerHistory().getTriggerId(),
                    data.getDataId());
        }

        context.incrementStat(numberOfDataEventsInserted,
                ChannelRouterContext.STAT_DATA_EVENTS_INSERTED);
        return numberOfDataEventsInserted;

    }

    protected int insertDataEvents(ChannelRouterContext context, DataMetaData dataMetaData,
            Collection<String> nodeIds, TriggerRouter triggerRouter) {
        int numberOfDataEventsInserted = 0;
        if (nodeIds == null || nodeIds.size() == 0) {
            nodeIds = new HashSet<String>(1);
            nodeIds.add(Constants.UNROUTED_NODE_ID);
        }
        long ts = System.currentTimeMillis();
        for (String nodeId : nodeIds) {
            Map<String, OutgoingBatch> batches = context.getBatchesByNodes();
            OutgoingBatch batch = batches.get(nodeId);
            if (batch == null) {
                batch = new OutgoingBatch(nodeId, dataMetaData.getNodeChannel().getChannelId(),
                        Status.RT);
                outgoingBatchService.insertOutgoingBatch(batch);
                context.getBatchesByNodes().put(nodeId, batch);
            }
            batch.incrementEventCount(dataMetaData.getData().getDataEventType());
            batch.incrementDataEventCount();
            numberOfDataEventsInserted++;
            context.addDataEvent(dataMetaData.getData().getDataId(), batch.getBatchId(),
                    triggerRouter.getRouter().getRouterId());
            if (batchAlgorithms.get(context.getChannel().getBatchAlgorithm()).isBatchComplete(
                    batch, dataMetaData, context)) {
                context.setNeedsCommitted(true);
            }
        }
        context.incrementStat(System.currentTimeMillis() - ts,
                ChannelRouterContext.STAT_INSERT_DATA_EVENTS_MS);
        return numberOfDataEventsInserted;
    }

    protected IDataRouter getDataRouter(TriggerRouter trigger) {
        IDataRouter router = null;
        if (!StringUtils.isBlank(trigger.getRouter().getRouterType())) {
            router = routers.get(trigger.getRouter().getRouterType());
            if (router == null) {
                log.warn("RouterMissing", trigger.getRouter().getRouterType(), trigger.getTrigger()
                        .getTriggerId());
            }
        }

        if (router == null) {
            return routers.get("default");
        }
        return router;
    }

    protected List<TriggerRouter> getTriggerRoutersForData(Data data) {
        List<TriggerRouter> triggerRouters = null;
        if (data != null) {
            if (data.getTriggerHistory() != null) {
                triggerRouters = triggerRouterService.getTriggerRoutersForCurrentNode(false).get(
                        (data.getTriggerHistory().getTriggerId()));
                if (triggerRouters == null || triggerRouters.size() == 0) {
                    triggerRouters = triggerRouterService.getTriggerRoutersForCurrentNode(true)
                            .get((data.getTriggerHistory().getTriggerId()));
                }
            } else {
                log.warn("TriggerHistMissing", data.getDataId());
            }
        }
        return triggerRouters;
    }

    public void addDataRouter(String name, IDataRouter dataRouter) {
        routers.put(name, dataRouter);
    }

    public void addBatchAlgorithm(String name, IBatchAlgorithm algorithm) {
        batchAlgorithms.put(name, algorithm);
    }

    public long getUnroutedDataCount() {
        long maxDataIdAlreadyRouted = sqlTemplate
                .queryForLong(getSql("selectLastDataIdRoutedUsingDataGapSql"));
        long leftToRoute = dataService.findMaxDataId() - maxDataIdAlreadyRouted;
        if (leftToRoute > 0) {
            return leftToRoute;
        } else {
            return 0;
        }
    }

    public List<String> getAvailableBatchAlgorithms() {
        return new ArrayList<String>(batchAlgorithms.keySet());
    }

    public Map<String, IDataRouter> getRouters() {
        return routers;
    }

    @Override
    protected AbstractSqlMap createSqlMap() {
        return new RouterServiceSqlMap(symmetricDialect.getPlatform(), createSqlReplacementTokens());
    }

}

File: symmetric-client-clib/src/service/RouterService.c
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "service/RouterService.h"
#include "common/Log.h"

static SymDataRouter * SymRouterService_getDataRouter(SymRouterService *this, SymRouter *router) {
    return this->routers->get(this->routers, SYM_ROUTER_DEFAULT);
}

static void SymRouterService_completeBatchesAndCommit(SymRouterService *this, SymChannelRouterContext *context) {
    SymList *usedRouters = context->usedDataRouters;
    SymList *batches = context->batchesByNodes->values(context->batchesByNodes);
    context->commit(context);

    SymIterator *iter = batches->iterator(batches);
    while (iter->hasNext(iter)) {
        SymOutgoingBatch *batch = (SymOutgoingBatch *) iter->next(iter);
        batch->routerMillis = (time(NULL) - batch->createTime->time) * 1000;
        SymIterator *routerIter = usedRouters->iterator(usedRouters);
        while (routerIter->hasNext(routerIter)) {
            SymDataRouter *router = (SymDataRouter *) routerIter->next(routerIter);
            router->completeBatch(router, context, batch);
        }
        routerIter->destroy(routerIter);
        if (SymStringUtils_equals(SYM_UNROUTED_NODE_ID, batch->nodeId)) {
            batch->status = SYM_OUTGOING_BATCH_OK;
        } else {
            batch->status = SYM_OUTGOING_BATCH_NEW;
        }
        this->outgoingBatchService->updateOutgoingBatch(this->outgoingBatchService, batch);
        context->batchesByNodes->remove(context->batchesByNodes, batch->nodeId);
    }
    iter->destroy(iter);

    SymIterator *routerIter = usedRouters->iterator(usedRouters);
    while (routerIter->hasNext(routerIter)) {
        SymDataRouter *router = (SymDataRouter *) routerIter->next(routerIter);
        router->contextCommitted(router, context);
    }
    routerIter->destroy(routerIter);
    context->needsCommitted = 0;
}

static int SymRouterService_insertDataEvents(SymRouterService *this, SymChannelRouterContext *context, SymDataMetaData *dataMetaData, SymList *nodeIds) {
    int numberOfDataEventsInserted = 0;
    if (nodeIds == NULL || nodeIds->size == 0) {
        nodeIds = SymList_new(NULL);
        nodeIds->add(nodeIds,SYM_UNROUTED_NODE_ID);
    }
    time_t ts = time(NULL);
    SymIterator *iter = nodeIds->iterator(nodeIds);
    while (iter->hasNext(iter)) {
        char *nodeId = (char *) iter->next(iter);
        if (nodeId) {
            SymMap *batches = context->batchesByNodes;
            SymOutgoingBatch *batch = batches->get(batches, nodeId);
            if (batch == NULL) {
                batch = SymOutgoingBatch_newWithNode(NULL, nodeId, dataMetaData->nodeChannel->channelId, SYM_OUTGOING_BATCH_ROUTING);

                SymLog_debug("About to insert a new batch for node %s on the '%s' channel.", nodeId, batch->channelId);

                this->outgoingBatchService->insertOutgoingBatch(this->outgoingBatchService, batch);
                context->batchesByNodes->put(context->batchesByNodes, nodeId, batch);
            }

            batch->incrementEventCount(batch, dataMetaData->data->eventType);
            batch->dataEventCount++;

            SymRouter *router = dataMetaData->router;
            context->addDataEvent(context, dataMetaData->data->dataId, batch->batchId, router ? router->routerId : SYM_UNKNOWN_ROUTER_ID);
            numberOfDataEventsInserted++;

            if (batch->dataEventCount >= dataMetaData->nodeChannel->maxBatchSize) {
                context->needsCommitted = 1;
            }
        }
    }
    context->statInsertDataEventsMs += ((time(NULL) - ts) * 1000);
    return numberOfDataEventsInserted;
}

SymList * SymRouterService_findAvailableNodes(SymRouterService *this, SymTriggerRouter *triggerRouter, SymChannelRouterContext *context) {
    SymList *nodes = context->availableNodes->get(context->availableNodes, triggerRouter->router->routerId);
    if (nodes == NULL) {
        nodes = SymList_new(NULL);
        SymRouter *router = triggerRouter->router;
        SymNodeGroupLink *link = this->configurationService->getNodeGroupLinkFor(this->configurationService,
                router->nodeGroupLink->sourceNodeGroupId, router->nodeGroupLink->targetNodeGroupId, 0);
        if (link != NULL) {
            nodes->addAll(nodes, this->nodeService->findEnabledNodesFromNodeGroup(this->nodeService, router->nodeGroupLink->targetNodeGroupId));
        } else {
            SymLog_error("The router %s has no node group link configured from %s to %s",
                    router->routerId, router->nodeGroupLink->sourceNodeGroupId, router->nodeGroupLink->targetNodeGroupId);
        }
        context->availableNodes->put(context->availableNodes, triggerRouter->router->routerId, nodes);
    }
    return nodes;
}

SymList * SymRouterService_getTriggerRoutersForData(SymRouterService *this, SymData *data) {
    SymList *triggerRouters = NULL;
    if (data != NULL) {
        if (data->triggerHistory) {
            SymMap *triggerRoutersMap = this->triggerRouterService->getTriggerRoutersForCurrentNode(this->triggerRouterService, 0);
            triggerRouters = triggerRoutersMap->get(triggerRoutersMap, data->triggerHistory->triggerId);
            if (triggerRouters == NULL || triggerRouters->size == 0) {
                triggerRoutersMap = this->triggerRouterService->getTriggerRoutersForCurrentNode(this->triggerRouterService, 1);
                triggerRouters = triggerRoutersMap->get(triggerRoutersMap, data->triggerHistory->triggerId);
            }
        } else {
            SymLog_warn("Could not find a trigger hist record for recorded data %ld.  Was the trigger hist record deleted manually?", data->dataId);
        }
    }
    return triggerRouters;
}

static int SymRouterService_routeDataToNodes(SymRouterService *this, SymData *data, SymChannelRouterContext *context) {
    int numberOfDataEventsInserted = 0;
    SymList *triggerRouters = SymRouterService_getTriggerRoutersForData(this, data);
    SymTable *table = this->platform->getTableFromCache(this->platform, data->triggerHistory->sourceCatalogName,
            data->triggerHistory->sourceSchemaName, data->triggerHistory->sourceTableName, 0);
    if (triggerRouters && triggerRouters->size > 0) {
        SymIterator *iter = triggerRouters->iterator(triggerRouters);
        while (iter->hasNext(iter)) {
            SymTriggerRouter *triggerRouter = (SymTriggerRouter *) iter->next(iter);

            SymDataMetaData *dataMetaData = SymDataMetaData_new(NULL, data, table, triggerRouter->router, context->channel);
            SymList *nodeIds = NULL;
            if (triggerRouter->isRouted(triggerRouter, data->eventType)) {
                SymDataRouter *dataRouter = SymRouterService_getDataRouter(this, triggerRouter->router);
                time_t ts = time(NULL);
                nodeIds = dataRouter->routeToNodes(dataRouter, context, dataMetaData,
                        SymRouterService_findAvailableNodes(this, triggerRouter, context), 0, 0, triggerRouter);
                context->statDataRouterMs += ((time(NULL) - ts) * 1000);

                if (nodeIds) {
                    if (!triggerRouter->pingBackEnabled && data->sourceNodeId) {
                        nodeIds->removeObject(nodeIds, data->sourceNodeId, (void *) strcmp);
                    }

                    // should never route to self
                    nodeIds->removeObject(nodeIds, context->nodeId, (void *) strcmp);
                }
            }
            numberOfDataEventsInserted += SymRouterService_insertDataEvents(this, context, dataMetaData, nodeIds);
            dataMetaData->destroy(dataMetaData);
        }
    } else {
        SymLog_warn("Could not find trigger routers for trigger history id of %d.  There is a good chance that data was captured and the trigger router link was removed before the data could be routed",
                data->triggerHistId);
        SymLog_info("Data with the id of %ld will be assigned to an unrouted batch", data->dataId);
        SymDataMetaData *dataMetaData = SymDataMetaData_new(NULL, data, table, NULL, context->channel);
        numberOfDataEventsInserted += SymRouterService_insertDataEvents(this, context, dataMetaData, NULL);
    }
    context->statDataEventsInserted += numberOfDataEventsInserted;
    return numberOfDataEventsInserted;
}

static int SymRouterService_selectDataAndRoute(SymRouterService *this, SymChannelRouterContext *context) {
    SymDataGapRouteReader *reader = SymDataGapRouteReader_new(NULL, this->platform, this->parameterService, this->dataService, context);
    int totalDataCount = 0;
    int totalDataEventCount = 0;
    int maxNumberOfEventsBeforeFlush = this->parameterService->getInt(this->parameterService, SYM_PARAMETER_ROUTING_FLUSH_JDBC_BATCH_SIZE, 50000);

    // TODO: use SymSqlCursor instead
    SymList *dataList = reader->selectDataFor(reader);
    SymIterator *iter = dataList->iterator(dataList);
    while (iter->hasNext(iter)) {
        SymData *data = (SymData *) iter->next(iter);
        totalDataCount++;
        int dataEventsInserted = SymRouterService_routeDataToNodes(this, data, context);
        totalDataEventCount += dataEventsInserted;

        if (maxNumberOfEventsBeforeFlush <= context->dataEventList->size || context->needsCommitted) {
            this->dataService->insertDataEvents(this->dataService, context->sqlTransaction, context->dataEventList);
            context->dataEventList->reset(context->dataEventList);
        }
        if (context->needsCommitted) {
            SymRouterService_completeBatchesAndCommit(this, context);
        }
    }
    context->statDataRoutedCount += totalDataCount;
    iter->destroy(iter);
    dataList->destroy(dataList);
    reader->destroy(reader);
    return totalDataEventCount;
}

static int SymRouterService_routeDataForChannel(SymRouterService *this, SymChannel *nodeChannel, SymNode *sourceNode) {
    SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);
    SymSqlTransaction *sqlTrans = sqlTemplate->startSqlTransaction(sqlTemplate);
    SymChannelRouterContext *context = SymChannelRouterContext_new(NULL, sourceNode->nodeId, nodeChannel,
            sqlTrans);
    int dataCount = SymRouterService_selectDataAndRoute(this, context);
    if (dataCount > 0) {
        long insertTs = time(NULL);
        this->dataService->insertDataEvents(this->dataService, context->sqlTransaction, context->dataEventList);
        SymRouterService_completeBatchesAndCommit(this, context);
        context->statInsertDataEventsMs += ((time(NULL) - insertTs) * 1000);
    }

    sqlTrans->destroy(sqlTrans);
    context->destroy(context);
    return dataCount;
}

static int SymRouterService_routeDataForEachChannel(SymRouterService *this) {
    int dataCount = 0;
    SymNode *sourceNode = this->nodeService->findIdentity(this->nodeService);
    SymMap *channels = this->configurationService->getChannels(this->configurationService, 0);
    SymList *channelList = channels->values(channels);
    SymIterator *iter = channelList->iterator(channelList);
    while (iter->hasNext(iter)) {
        SymChannel *nodeChannel = (SymChannel *) iter->next(iter);
        if (nodeChannel->enabled) {
            dataCount += SymRouterService_routeDataForChannel(this, nodeChannel, sourceNode);
        } else {
            SymLog_debug("Not routing the %s channel.  It is either disabled or suspended.", nodeChannel->channelId);
        }
    }
    iter->destroy(iter);
    channelList->destroy(channelList);
    channels->destroyAll(channels, (void *)SymChannel_destroy);
    return dataCount;
}

long SymRouterService_routeData(SymRouterService *this) {
	time_t ts = time(NULL);
	int dataCount = SymRouterService_routeDataForEachChannel(this);
    ts = time(NULL) - ts;
    if (dataCount > 0 || ts * 1000 > SYM_LONG_OPERATION_THRESHOLD) {
        SymLog_info("Routed %d data events in %d ms", dataCount, ts);
    }
    return dataCount;
}

void SymRouterService_destroy(SymRouterService *this) {
    this->routers->destroyAll(this->routers, (void *) SymRouter_destroy);
    free(this);
}

SymRouterService * SymRouterService_new(SymRouterService *this, SymOutgoingBatchService *outgoingBatchService, SymSequenceService *sequenceService,
        SymDataService *dataService, SymNodeService *nodeService, SymConfigurationService *configurationService, SymParameterService *parameterService,
        SymTriggerRouterService *triggerRouterService, SymDatabasePlatform *platform) {
    if (this == NULL) {
        this = (SymRouterService *) calloc(1, sizeof(SymRouterService));
    }
    this->outgoingBatchService = outgoingBatchService;
    this->sequenceService = sequenceService;
    this->dataService = dataService;
    this->nodeService = nodeService;
    this->configurationService = configurationService;
    this->parameterService = parameterService;
    this->triggerRouterService = triggerRouterService;
    this->platform = platform;
    this->routers = SymMap_new(NULL, 20);
    this->routers->put(this->routers, SYM_ROUTER_DEFAULT, SymDefaultDataRouter_new(NULL));
    this->routeData = (void *) &SymRouterService_routeData;
    this->destroy = (void *) &SymRouterService_destroy;
    return this;
}
